src/coreclr/src/tools/Common/JitInterface/ThunkGenerator/Program.cs

```diff
@@ -249,80 +264,23 @@ unsafe partial class CorInfoImpl
-            }
-            tr.WriteLine();
-#endif
-
-            foreach (FunctionDecl decl in functionData)
-            {
-                tr.WriteLine("        [UnmanagedFunctionPointerAttribute(default(CallingConvention))]");
-
-                string returnType = decl.ReturnAsParm ? "void" : decl.ReturnType.ManagedTypeName;
-                int marshalAs = returnType.LastIndexOf(']');
-                string returnTypeWithDelegate = returnType.Insert((marshalAs != -1) ? (marshalAs + 1) : 0, "delegate ");
-
-                tr.Write("        " + returnTypeWithDelegate + " " + "__" + decl.FunctionName + "(");
-                tr.Write("IntPtr _this");
-                tr.Write(", IntPtr* ppException");
-                if (decl.ReturnAsParm)
-                {
-                    tr.Write(", out " + decl.ReturnType.ManagedTypeName + " _return");
-                }
-                foreach (Parameter param in decl.Parameters)
-                {
-                    tr.Write(", ");
-                    tr.Write(param.Type.ManagedTypeName + " " + param.Name);
-                }
-                tr.WriteLine(");");
-            }
-            tr.WriteLine();
-
             foreach (FunctionDecl decl in functionData)
             {
-                string returnType = decl.ReturnAsParm ? "void" : decl.ReturnType.ManagedTypeName;
-                int marshalAs = returnType.LastIndexOf(']');
-                string returnTypeWithStatic = returnType.Insert((marshalAs != -1) ? (marshalAs + 1) : 0, "static ");
-
-                tr.Write("        " + returnTypeWithStatic + " " + "_" + decl.FunctionName + "(");
-                tr.Write("IntPtr thisHandle");
-                tr.Write(", IntPtr* ppException");
-                if (decl.ReturnAsParm)
-                {
-                    tr.Write(", out " + decl.ReturnType.ManagedTypeName + " _return");
-                }
+                tr.WriteLine("        [UnmanagedCallersOnly]");
+                tr.Write($"        static {decl.ReturnType.UnmanagedTypeName} _{decl.FunctionName}(IntPtr thisHandle, IntPtr* ppException");
                 foreach (Parameter param in decl.Parameters)
                 {
-                    tr.Write(", ");
-                    tr.Write(param.Type.ManagedTypeName + " " + param.Name);
+                    tr.Write($", {param.Type.UnmanagedTypeName} {param.Name}");
                 }
                 tr.Write(@")
         {
             var _this = GetThis(thisHandle);
             try
             {
 ");
-                bool isVoid = decl.ReturnAsParm || decl.ReturnType.ManagedTypeName == "void";
-                tr.Write("                " + (isVoid ? "" : "return ") + "_this." + decl.FunctionName + "(");
+                bool isVoid = decl.ReturnType.ManagedTypeName == "void";
+                tr.Write($"                " + (isVoid ? "" : "return ") + $"_this.{decl.FunctionName}(");
```

Comment:
nit: fold in single interpolation
```suggestion
                tr.Write($"                {(isVoid ? "" : "return ")}_this.{decl.FunctionName}(");
```